File: lang/python/core/src/lws/cli/services/sns.py
"""``lws sns`` sub-commands."""

from __future__ import annotations

import asyncio
import json

import typer

from lws.cli.services.client import LwsClient, exit_with_error, output_json, xml_to_dict

app = typer.Typer(help="SNS commands")

_SERVICE = "sns"


def _client(port: int) -> LwsClient:
    return LwsClient(port=port)


@app.command("publish")
def publish(
    topic_name: str = typer.Option(..., "--topic-name", help="Topic name"),
    message: str = typer.Option(..., "--message", help="Message body"),
    subject: str = typer.Option(None, "--subject", help="Message subject"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """Publish a message to a topic."""
    asyncio.run(_publish(topic_name, message, subject, port))


async def _publish(topic_name: str, message: str, subject: str | None, port: int) -> None:
    client = _client(port)
    try:
        resource = await client.resolve_resource(_SERVICE, topic_name)
        topic_arn = resource.get("arn", f"arn:aws:sns:us-east-1:000000000000:{topic_name}")
    except Exception:
        topic_arn = f"arn:aws:sns:us-east-1:000000000000:{topic_name}"
    params: dict[str, str] = {
        "Action": "Publish",
        "TopicArn": topic_arn,
        "Message": message,
    }
    if subject:
        params["Subject"] = subject
    xml = await client.form_request(_SERVICE, params)
    output_json(xml_to_dict(xml))


@app.command("list-topics")
def list_topics(
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """List all topics."""
    asyncio.run(_list_topics(port))


async def _list_topics(port: int) -> None:
    client = _client(port)
    try:
        xml = await client.form_request(_SERVICE, {"Action": "ListTopics"})
    except Exception as exc:
        exit_with_error(str(exc))
    output_json(xml_to_dict(xml))


@app.command("list-subscriptions")
def list_subscriptions(
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """List all subscriptions."""
    asyncio.run(_list_subscriptions(port))


async def _list_subscriptions(port: int) -> None:
    client = _client(port)
    try:
        xml = await client.form_request(_SERVICE, {"Action": "ListSubscriptions"})
    except Exception as exc:
        exit_with_error(str(exc))
    output_json(xml_to_dict(xml))


@app.command("create-topic")
def create_topic(
    name: str = typer.Option(..., "--name", help="Topic name"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """Create a topic."""
    asyncio.run(_create_topic(name, port))


async def _create_topic(name: str, port: int) -> None:
    client = _client(port)
    try:
        xml = await client.form_request(
            _SERVICE,
            {"Action": "CreateTopic", "Name": name},
        )
    except Exception as exc:
        exit_with_error(str(exc))
    output_json(xml_to_dict(xml))


@app.command("delete-topic")
def delete_topic(
    topic_arn: str = typer.Option(..., "--topic-arn", help="Topic ARN"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """Delete a topic."""
    asyncio.run(_delete_topic(topic_arn, port))


async def _delete_topic(topic_arn: str, port: int) -> None:
    client = _client(port)
    try:
        xml = await client.form_request(
            _SERVICE,
            {"Action": "DeleteTopic", "TopicArn": topic_arn},
        )
    except Exception as exc:
        exit_with_error(str(exc))
    output_json(xml_to_dict(xml))


@app.command("subscribe")
def subscribe(
    topic_arn: str = typer.Option(..., "--topic-arn", help="Topic ARN"),
    protocol: str = typer.Option(..., "--protocol", help="Subscription protocol"),
    endpoint: str = typer.Option(..., "--notification-endpoint", help="Subscription endpoint"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """Subscribe an endpoint to a topic."""
    asyncio.run(_subscribe(topic_arn, protocol, endpoint, port))


async def _subscribe(topic_arn: str, protocol: str, endpoint: str, port: int) -> None:
    client = _client(port)
    try:
        xml = await client.form_request(
            _SERVICE,
            {
                "Action": "Subscribe",
                "TopicArn": topic_arn,
                "Protocol": protocol,
                "Endpoint": endpoint,
            },
        )
    except Exception as exc:
        exit_with_error(str(exc))
    output_json(xml_to_dict(xml))


@app.command("unsubscribe")
def unsubscribe(
    subscription_arn: str = typer.Option(..., "--subscription-arn", help="Subscription ARN"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """Unsubscribe from a topic."""
    asyncio.run(_unsubscribe(subscription_arn, port))


async def _unsubscribe(subscription_arn: str, port: int) -> None:
    client = _client(port)
    xml = await client.form_request(
        _SERVICE,
        {"Action": "Unsubscribe", "SubscriptionArn": subscription_arn},
    )
    output_json(xml_to_dict(xml))


@app.command("get-topic-attributes")
def get_topic_attributes(
    topic_arn: str = typer.Option(..., "--topic-arn", help="Topic ARN"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """Get topic attributes."""
    asyncio.run(_get_topic_attributes(topic_arn, port))


async def _get_topic_attributes(topic_arn: str, port: int) -> None:
    client = _client(port)
    xml = await client.form_request(
        _SERVICE,
        {"Action": "GetTopicAttributes", "TopicArn": topic_arn},
    )
    output_json(xml_to_dict(xml))


@app.command("set-topic-attributes")
def set_topic_attributes(
    topic_arn: str = typer.Option(..., "--topic-arn", help="Topic ARN"),
    attribute_name: str = typer.Option(..., "--attribute-name", help="Attribute name"),
    attribute_value: str = typer.Option(..., "--attribute-value", help="Attribute value"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """Set a topic attribute."""
    asyncio.run(_set_topic_attributes(topic_arn, attribute_name, attribute_value, port))


async def _set_topic_attributes(
    topic_arn: str, attribute_name: str, attribute_value: str, port: int
) -> None:
    client = _client(port)
    xml = await client.form_request(
        _SERVICE,
        {
            "Action": "SetTopicAttributes",
            "TopicArn": topic_arn,
            "AttributeName": attribute_name,
            "AttributeValue": attribute_value,
        },
    )
    output_json(xml_to_dict(xml))


@app.command("list-tags-for-resource")
def list_tags_for_resource(
    resource_arn: str = typer.Option(..., "--resource-arn", help="Resource ARN"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """List tags for a resource."""
    asyncio.run(_list_tags_for_resource(resource_arn, port))


async def _list_tags_for_resource(resource_arn: str, port: int) -> None:
    client = _client(port)
    xml = await client.form_request(
        _SERVICE,
        {"Action": "ListTagsForResource", "ResourceArn": resource_arn},
    )
    output_json(xml_to_dict(xml))


@app.command("tag-resource")
def tag_resource(
    resource_arn: str = typer.Option(..., "--resource-arn", help="Resource ARN"),
    tags: str = typer.Option(..., "--tags", help="JSON array of tag objects"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """Tag a resource."""
    asyncio.run(_tag_resource(resource_arn, tags, port))


async def _tag_resource(resource_arn: str, tags: str, port: int) -> None:
    client = _client(port)
    params: dict[str, str] = {
        "Action": "TagResource",
        "ResourceArn": resource_arn,
    }
    tag_list = json.loads(tags)
    for i, tag in enumerate(tag_list, start=1):
        params[f"Tags.member.{i}.Key"] = tag["Key"]
        params[f"Tags.member.{i}.Value"] = tag["Value"]
    xml = await client.form_request(_SERVICE, params)
    output_json(xml_to_dict(xml))


@app.command("untag-resource")
def untag_resource(
    resource_arn: str = typer.Option(..., "--resource-arn", help="Resource ARN"),
    tag_keys: str = typer.Option(..., "--tag-keys", help="JSON array of tag keys"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """Remove tags from a resource."""
    asyncio.run(_untag_resource(resource_arn, tag_keys, port))


async def _untag_resource(resource_arn: str, tag_keys: str, port: int) -> None:
    client = _client(port)
    params: dict[str, str] = {
        "Action": "UntagResource",
        "ResourceArn": resource_arn,
    }
    keys = json.loads(tag_keys)
    for i, key in enumerate(keys, start=1):
        params[f"TagKeys.member.{i}"] = key
    xml = await client.form_request(_SERVICE, params)
    output_json(xml_to_dict(xml))


@app.command("get-subscription-attributes")
def get_subscription_attributes(
    subscription_arn: str = typer.Option(..., "--subscription-arn", help="Subscription ARN"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """Get subscription attributes."""
    asyncio.run(_get_subscription_attributes(subscription_arn, port))


async def _get_subscription_attributes(subscription_arn: str, port: int) -> None:
    client = _client(port)
    xml = await client.form_request(
        _SERVICE,
        {"Action": "GetSubscriptionAttributes", "SubscriptionArn": subscription_arn},
    )
    output_json(xml_to_dict(xml))


@app.command("set-subscription-attributes")
def set_subscription_attributes(
    subscription_arn: str = typer.Option(..., "--subscription-arn", help="Subscription ARN"),
    attribute_name: str = typer.Option(..., "--attribute-name", help="Attribute name"),
    attribute_value: str = typer.Option(..., "--attribute-value", help="Attribute value"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """Set a subscription attribute."""
    asyncio.run(
        _set_subscription_attributes(subscription_arn, attribute_name, attribute_value, port)
    )


async def _set_subscription_attributes(
    subscription_arn: str, attribute_name: str, attribute_value: str, port: int
) -> None:
    client = _client(port)
    xml = await client.form_request(
        _SERVICE,
        {
            "Action": "SetSubscriptionAttributes",
            "SubscriptionArn": subscription_arn,
            "AttributeName": attribute_name,
            "AttributeValue": attribute_value,
        },
    )
    output_json(xml_to_dict(xml))


@app.command("confirm-subscription")
def confirm_subscription(
    topic_arn: str = typer.Option(..., "--topic-arn", help="Topic ARN"),
    token: str = typer.Option(..., "--token", help="Subscription confirmation token"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """Confirm a subscription."""
    asyncio.run(_confirm_subscription(topic_arn, token, port))


async def _confirm_subscription(topic_arn: str, token: str, port: int) -> None:
    client = _client(port)
    xml = await client.form_request(
        _SERVICE,
        {"Action": "ConfirmSubscription", "TopicArn": topic_arn, "Token": token},
    )
    output_json(xml_to_dict(xml))


@app.command("list-subscriptions-by-topic")
def list_subscriptions_by_topic(
    topic_arn: str = typer.Option(..., "--topic-arn", help="Topic ARN"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """List subscriptions for a topic."""
    asyncio.run(_list_subscriptions_by_topic(topic_arn, port))


async def _list_subscriptions_by_topic(topic_arn: str, port: int) -> None:
    client = _client(port)
    xml = await client.form_request(
        _SERVICE,
        {"Action": "ListSubscriptionsByTopic", "TopicArn": topic_arn},
    )
    output_json(xml_to_dict(xml))

File: lang/typescript/core/tests/steps/sns.ts
/** SNS step definitions. */

import { Given, When, Then } from "@cucumber/cucumber";
import assert from "assert";
import {
  CreateTopicCommand,
  DeleteTopicCommand,
  ListTopicsCommand,
  PublishCommand,
  SubscribeCommand,
  UnsubscribeCommand,
  ListSubscriptionsCommand,
  ListSubscriptionsByTopicCommand,
  GetTopicAttributesCommand,
  SetTopicAttributesCommand,
  GetSubscriptionAttributesCommand,
  SetSubscriptionAttributesCommand,
  TagResourceCommand,
  UntagResourceCommand,
  ListTagsForResourceCommand,
  ConfirmSubscriptionCommand,
} from "@aws-sdk/client-sns";
import type { LwsWorld } from "../support/world";

const ACCOUNT = "000000000000";
const REGION = "us-east-1";

function topicArn(port: number, topicName: string): string {
  return `arn:aws:sns:${REGION}:${ACCOUNT}:${topicName}`;
}

async function createTopic(world: LwsWorld, topicName: string): Promise<string> {
  const client = world.snsClient();
  const result = await client.send(new CreateTopicCommand({ Name: topicName }));
  return result.TopicArn!;
}

// --- Given -----------------------------------------------------------------

Given("a topic {string} was created", async function (this: LwsWorld, topicName: string) {
  const arn = await createTopic(this, topicName);
  this.lastTopicArn = arn;
});

Given("an SQS subscription to {string} was added to topic {string}", async function (
  this: LwsWorld,
  sqsArn: string,
  topicName: string
) {
  const client = this.snsClient();
  const arn = topicArn(this.snsPort, topicName);
  const result = await client.send(
    new SubscribeCommand({ TopicArn: arn, Protocol: "sqs", Endpoint: sqsArn })
  );
  this.lastSubscriptionArn = result.SubscriptionArn;
});

Given("tags {string} were added to topic {string}", async function (
  this: LwsWorld,
  tagsJson: string,
  topicName: string
) {
  const client = this.snsClient();
  const arn = topicArn(this.snsPort, topicName);
  const tags = (JSON.parse(tagsJson) as Array<{ Key: string; Value: string }>).map((t) => ({
    Key: t.Key,
    Value: t.Value,
  }));
  await client.send(new TagResourceCommand({ ResourceArn: arn, Tags: tags }));
});

// --- When ------------------------------------------------------------------

When("I create topic {string}", async function (this: LwsWorld, topicName: string) {
  const client = this.snsClient();
  try {
    const result = await client.send(new CreateTopicCommand({ Name: topicName }));
    this.lastTopicArn = result.TopicArn;
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I delete the topic {string}", async function (this: LwsWorld, topicName: string) {
  const client = this.snsClient();
  const arn = topicArn(this.snsPort, topicName);
  try {
    const result = await client.send(new DeleteTopicCommand({ TopicArn: arn }));
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I list topics", async function (this: LwsWorld) {
  const client = this.snsClient();
  try {
    const result = await client.send(new ListTopicsCommand({}));
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I publish message {string} to topic {string}", async function (
  this: LwsWorld,
  message: string,
  topicName: string
) {
  const client = this.snsClient();
  const arn = topicArn(this.snsPort, topicName);
  try {
    const result = await client.send(
      new PublishCommand({ TopicArn: arn, Message: message })
    );
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I subscribe {string} with protocol {string} to the topic {string}", async function (
  this: LwsWorld,
  endpoint: string,
  protocol: string,
  topicName: string
) {
  const client = this.snsClient();
  const arn = topicArn(this.snsPort, topicName);
  try {
    const result = await client.send(
      new SubscribeCommand({ TopicArn: arn, Protocol: protocol, Endpoint: endpoint })
    );
    this.lastSubscriptionArn = result.SubscriptionArn;
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I unsubscribe the subscription on topic {string}", async function (
  this: LwsWorld,
  _topicName: string
) {
  const client = this.snsClient();
  try {
    const result = await client.send(
      new UnsubscribeCommand({ SubscriptionArn: this.lastSubscriptionArn! })
    );
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I list subscriptions", async function (this: LwsWorld) {
  const client = this.snsClient();
  try {
    const result = await client.send(new ListSubscriptionsCommand({}));
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I list subscriptions by topic {string}", async function (
  this: LwsWorld,
  topicName: string
) {
  const client = this.snsClient();
  const arn = topicArn(this.snsPort, topicName);
  try {
    const result = await client.send(new ListSubscriptionsByTopicCommand({ TopicArn: arn }));
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I get topic attributes for topic {string}", async function (
  this: LwsWorld,
  topicName: string
) {
  const client = this.snsClient();
  const arn = topicArn(this.snsPort, topicName);
  try {
    const result = await client.send(new GetTopicAttributesCommand({ TopicArn: arn }));
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I set topic attribute {string} to {string} for topic {string}", async function (
  this: LwsWorld,
  attrName: string,
  attrValue: string,
  topicName: string
) {
  const client = this.snsClient();
  const arn = topicArn(this.snsPort, topicName);
  try {
    const result = await client.send(
      new SetTopicAttributesCommand({ TopicArn: arn, AttributeName: attrName, AttributeValue: attrValue })
    );
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I get subscription attributes for the subscription on topic {string}", async function (
  this: LwsWorld,
  _topicName: string
) {
  const client = this.snsClient();
  try {
    const result = await client.send(
      new GetSubscriptionAttributesCommand({ SubscriptionArn: this.lastSubscriptionArn! })
    );
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I set subscription attribute {string} to {string} for the subscription on topic {string}", async function (
  this: LwsWorld,
  attrName: string,
  attrValue: string,
  _topicName: string
) {
  const client = this.snsClient();
  try {
    const result = await client.send(
      new SetSubscriptionAttributesCommand({
        SubscriptionArn: this.lastSubscriptionArn!,
        AttributeName: attrName,
        AttributeValue: attrValue,
      })
    );
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I tag resource {string} with tags {string}", async function (
  this: LwsWorld,
  topicName: string,
  tagsJson: string
) {
  const client = this.snsClient();
  const arn = topicArn(this.snsPort, topicName);
  const tags = (JSON.parse(tagsJson) as Array<{ Key: string; Value: string }>).map((t) => ({
    Key: t.Key,
    Value: t.Value,
  }));
  try {
    const result = await client.send(new TagResourceCommand({ ResourceArn: arn, Tags: tags }));
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I untag resource {string} with tag keys {string}", async function (
  this: LwsWorld,
  topicName: string,
  tagKeysJson: string
) {
  const client = this.snsClient();
  const arn = topicArn(this.snsPort, topicName);
  const tagKeys = JSON.parse(tagKeysJson) as string[];
  try {
    const result = await client.send(new UntagResourceCommand({ ResourceArn: arn, TagKeys: tagKeys }));
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I list tags for resource {string}", async function (this: LwsWorld, resourceRef: string) {
  // Route based on whether it's an EventBridge ARN or SNS topic name
  if (resourceRef.startsWith("arn:aws:events:")) {
    const { EventBridgeClient, ListTagsForResourceCommand: EbListTags } = await import("@aws-sdk/client-eventbridge");
    const client = new EventBridgeClient({
      endpoint: `http://127.0.0.1:${this.eventbridgePort}`,
      credentials: { accessKeyId: "test", secretAccessKey: "test" },
      region: "us-east-1",
    });
    try {
      const result = await client.send(new EbListTags({ ResourceARN: resourceRef }));
      this.lastResult = { success: true, output: result };
    } catch (err) {
      this.lastResult = { success: false, output: err, error: err };
    }
  } else {
    const client = this.snsClient();
    const arn = topicArn(this.snsPort, resourceRef);
    try {
      const result = await client.send(new ListTagsForResourceCommand({ ResourceArn: arn }));
      this.lastResult = { success: true, output: result };
    } catch (err) {
      this.lastResult = { success: false, output: err, error: err };
    }
  }
});

When("I confirm subscription for topic {string} with token {string}", async function (
  this: LwsWorld,
  topicName: string,
  token: string
) {
  const client = this.snsClient();
  const arn = topicArn(this.snsPort, topicName);
  try {
    const result = await client.send(
      new ConfirmSubscriptionCommand({ TopicArn: arn, Token: token })
    );
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

// Timed variant
When("I list SNS topics with timing", async function (this: LwsWorld) {
  const client = this.snsClient();
  const start = Date.now();
  try {
    const result = await client.send(new ListTopicsCommand({}));
    this.timedResult = { success: true, output: result, elapsedMs: Date.now() - start };
  } catch (err) {
    this.timedResult = { success: false, output: err, elapsedMs: Date.now() - start };
  }
});

When("I list SNS topics", async function (this: LwsWorld) {
  const client = this.snsClient();
  try {
    const result = await client.send(new ListTopicsCommand({}));
    this.lastResult = { success: true, output: result };
  } catch (err: unknown) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

// --- Then ------------------------------------------------------------------

Then("the topic {string} will appear in the topic list", async function (
  this: LwsWorld,
  topicName: string
) {
  const client = this.snsClient();
  const result = await client.send(new ListTopicsCommand({}));
  const arns = (result.Topics ?? []).map((t) => t.TopicArn ?? "");
  assert.ok(
    arns.some((a) => a.includes(topicName)),
    `Expected topic "${topicName}" in list but got: ${arns.join(", ")}`
  );
});

Then("the topic {string} will not appear in the topic list", async function (
  this: LwsWorld,
  topicName: string
) {
  const client = this.snsClient();
  const result = await client.send(new ListTopicsCommand({}));
  const arns = (result.Topics ?? []).map((t) => t.TopicArn ?? "");
  assert.ok(
    !arns.some((a) => a.includes(topicName)),
    `Expected topic "${topicName}" to not be in list`
  );
});

Then("the output will contain a ListSubscriptionsResponse", function (this: LwsWorld) {
  const output = this.lastResult.output as { Subscriptions?: unknown[] };
  assert.ok(output?.Subscriptions !== undefined, "Expected Subscriptions in output");
});

Then("the output will contain a PublishResponse", function (this: LwsWorld) {
  const output = this.lastResult.output as { MessageId?: string };
  assert.ok(output?.MessageId, "Expected MessageId in PublishResponse");
});

Then("the topic {string} will have a subscription in the subscription list", async function (
  this: LwsWorld,
  topicName: string
) {
  const client = this.snsClient();
  const arn = topicArn(this.snsPort, topicName);
  const result = await client.send(new ListSubscriptionsByTopicCommand({ TopicArn: arn }));
  assert.ok((result.Subscriptions ?? []).length > 0, "Expected at least one subscription");
});

Then("the topic {string} will not have a subscription in the subscription list", async function (
  this: LwsWorld,
  topicName: string
) {
  const client = this.snsClient();
  const arn = topicArn(this.snsPort, topicName);
  const result = await client.send(new ListSubscriptionsByTopicCommand({ TopicArn: arn }));
  assert.strictEqual((result.Subscriptions ?? []).length, 0, "Expected no subscriptions");
});

Then("the topic {string} will have display name {string}", async function (
  this: LwsWorld,
  topicName: string,
  expectedDisplayName: string
) {
  const client = this.snsClient();
  const arn = topicArn(this.snsPort, topicName);
  const result = await client.send(new GetTopicAttributesCommand({ TopicArn: arn }));
  const actualDisplayName = result.Attributes?.DisplayName;
  assert.strictEqual(actualDisplayName, expectedDisplayName);
});

Then("the subscription on topic {string} will have attribute {string} equal to {string}", async function (
  this: LwsWorld,
  _topicName: string,
  attrName: string,
  expectedValue: string
) {
  const client = this.snsClient();
  const result = await client.send(
    new GetSubscriptionAttributesCommand({ SubscriptionArn: this.lastSubscriptionArn! })
  );
  const actualValue = result.Attributes?.[attrName];
  assert.strictEqual(actualValue, expectedValue);
});

Then("topic {string} will have tag {string} with value {string}", async function (
  this: LwsWorld,
  topicName: string,
  tagKey: string,
  expectedValue: string
) {
  const client = this.snsClient();
  const arn = topicArn(this.snsPort, topicName);
  const result = await client.send(new ListTagsForResourceCommand({ ResourceArn: arn }));
  const tags = result.Tags ?? [];
  const tag = tags.find((t) => t.Key === tagKey);
  assert.strictEqual(tag?.Value, expectedValue);
});

Then("topic {string} will not have tag {string}", async function (
  this: LwsWorld,
  topicName: string,
  tagKey: string
) {
  const client = this.snsClient();
  const arn = topicArn(this.snsPort, topicName);
  const result = await client.send(new ListTagsForResourceCommand({ ResourceArn: arn }));
  const tags = result.Tags ?? [];
  const tag = tags.find((t) => t.Key === tagKey);
  assert.ok(!tag, `Expected tag "${tagKey}" to not exist`);
});
